sqlglot/dialects.py

```diff
@@ -206,33 +206,37 @@ def ts_or_ds_to_date(self, expression):
     }
 
     functions = {
-        'APPROX_COUNT_DISTINCT': lambda args: exp.ApproxDistinct(this=args[0]),
-        'COLLECT_LIST': lambda args: exp.ArrayAgg(this=args[0]),
-        'DATE_ADD': lambda args: exp.DateAdd(this=exp.DateStrToDate(this=args[0]), expression=args[1]),
-        'DATEDIFF': lambda args: exp.DateDiff(
-            this=exp.DateStrToDate(this=args[0]),
-            expression=exp.DateStrToDate(this=args[1]),
-        ),
-        'DATE_SUB': lambda args: exp.DateAdd(
-            this=exp.DateStrToDate(this=args[0]),
-            expression=exp.Star(this=args[1], expression=Token.number(-1)),
-        ),
-        'DATE_FORMAT': lambda args: exp.TimeToStr(this=args[0], format=args[1]),
-        'DAY': lambda args: exp.Day(this=exp.TsOrDsToDate(this=args[0])),
-        'FROM_UNIXTIME': lambda args: exp.UnixToStr(
-            this=args[0],
-            format=list_get(args, 1) or Hive.TIME_FORMAT,
-        ),
-        'GET_JSON_OBJECT': lambda args: exp.JSONPath(this=args[0], path=args[1]),
-        'LOCATE': lambda args: exp.StrPosition(this=args[1], substr=args[0], position=list_get(args, 2)),
-        'MONTH': lambda args: exp.Month(this=exp.TsOrDsToDate(this=args[0])),
-        'PERCENTILE': lambda args: exp.Quantile(this=args[0], quantile=args[1]),
-        'SIZE': lambda args: exp.ArraySize(this=args[0]),
-        'TO_DATE': lambda args: exp.TsOrDsToDateStr(this=args[0]),
-        'UNIX_TIMESTAMP': lambda args: exp.StrToUnix(
-            this=args[0],
-            format=list_get(args, 1) or Hive.TIME_FORMAT,
-        ),
+        'APPROX_COUNT_DISTINCT': exp.ApproxDistinct.from_arg_list,
+        'COLLECT_LIST': exp.ArrayAgg.from_arg_list,
+        'DATE_ADD': lambda args: exp.DateAdd.from_arg_list([exp.DateStrToDate(this=list_get(args, 0)), *args[1:]]),
+        'DATEDIFF': lambda args: exp.DateDiff.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
+            exp.DateStrToDate(this=list_get(args, 1)),
+            *args[2:]
+        ]),
+        'DATE_SUB': lambda args: exp.DateAdd.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
```

Comment:
is from_arg_list really more clear?

Reply:
what do you think about for any default version, we use from_arg_list, but if you have to change order then do it the old way?

Reply:
That's a good question and I was thinking about this myself. I concluded that the `from_arg_list` is still a better option since it protects against excessive arguments out of the box. What do you think?

Reply:
with my snippet below, i don't think it's necessary anymore for the excessive argument check